src/libraries/System.Private.CoreLib/src/System/Type.cs

```diff
@@ -119,10 +119,20 @@ public virtual Type[] GetGenericParameterConstraints()
         public virtual bool IsSecurityTransparent => throw NotImplemented.ByDesign;
 
         public virtual StructLayoutAttribute? StructLayoutAttribute => throw new NotSupportedException();
-        public ConstructorInfo? TypeInitializer => GetConstructorImpl(BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic, null, CallingConventions.Any, Type.EmptyTypes, null);
 
+        public ConstructorInfo? TypeInitializer
+        {
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)]
+            get => GetConstructorImpl(BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic, null, CallingConventions.Any, Type.EmptyTypes, null);
```

Comment:
This only cares about static cctors, but the DynamicallyAccessedMembers will keep all constructors alive.  Do we care?  Do we need a more fine-grained DynamicallyAccessedMemberTypes?

Reply:
I was considering both static/instance divide and public/nonpublic divide. Public/nonpublic matches most reflection APIs so it was the more natural choice.

This is one of the places APIs where static/instance would help...

We could potentially intrinisically recognize this property in the linker and if the type is statically known, only keep the `.cctor`. But based on [API usage](https://apisof.net/catalog/System.Type.TypeInitializer) it doesn't look worth the effort since nobody uses this API.